File: java/components/StrataCommon/src/strata1/common/producerconsumer/BlockingQueueDispatcher.java
// ##########################################################################
// # File Name:	BlockingQueueDispatcher.java
// #
// # Copyright:	2014, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataCommon Framework.
// #
// #   			The StrataCommon Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataCommon Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataCommon
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata1.common.producerconsumer;

import strata1.common.utility.IMultiMap;
import strata1.common.utility.MultiMap;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.PriorityBlockingQueue;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public
class BlockingQueueDispatcher<T>
    extends    AbstractPriorityDispatcher<T>
    implements IPriorityDispatcher<T>
{
    private ExecutorService                                 itsPool;
    private IMultiMap<ISelector<T>,BlockingQueue<Entry<T>>> itsQueues;
    private AtomicBoolean                                   itsRunningFlag;
    private AtomicLong                                      itsSequence;
    
    /************************************************************************
     * Creates a new {@code BlockingQueueDispatcher}. 
     *
     */
    protected 
    BlockingQueueDispatcher()
    {
        itsPool   = null;
        itsQueues = new MultiMap<ISelector<T>,BlockingQueue<Entry<T>>>();
        itsRunningFlag = new AtomicBoolean(false);
        itsSequence = new AtomicLong(0L);
    }
    
    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void
    route(int priority,T payload)
    {
        long sequence = 0L;
        
        if ( priority < 0 )
            throw 
                new IllegalArgumentException("priority cannot be negative.");
        
        if ( !isDispatching() )
            throw new IllegalStateException("Not dispatching.");
        
        sequence = itsSequence.getAndIncrement();
        
        for (ISelector<T> selector:itsQueues.getKeys())
        {
            if ( selector.match( payload ) )
            {
                BlockingQueue<Entry<T>> queue = 
                    getQueue( selector,sequence );
                
                try
                {
                    queue.put( new Entry<T>(priority,payload ) );
                }
                catch(InterruptedException e)
                {
                }
            }
        } 
    }
    
    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void
    broadcast(int priority,T payload)
    {
        if ( priority < 0 )
            throw 
                new IllegalArgumentException("priority cannot be negative.");

        if ( !isDispatching() )
            throw new IllegalStateException("Not dispatching.");
        
        for (ISelector<T> selector:itsQueues.getKeys())
        {
            if ( selector.match( payload ) )
            {
                for (BlockingQueue<Entry<T>> queue:itsQueues.get( selector ) )
                {
                    try
                    {
                        queue.put( new Entry<T>(priority,payload ) );
                    }
                    catch(InterruptedException e)
                    {
                    }
                }
            }
        }        
    }
    
    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    startDispatching()
    {
        if ( !isDispatching() )
        {
            getRunningFlag().compareAndSet( false,true );
            setExecutor( Executors.newCachedThreadPool() );
           
            for (ISelector<T> selector:getConsumers().getKeys())                
            {
                for (IConsumer<T> consumer:getConsumers().get( selector ))
                {
                    BlockingQueue<Entry<T>> queue = 
                        new PriorityBlockingQueue<Entry<T>>();
                    
                    addToQueues( selector,queue );
                    getExecutor().execute( 
                        new BlockingQueueProcessor<T>(
                            queue,
                            consumer,
                            getRunningFlag()) );
                }
            }
        }
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    stopDispatching()
    {
        if ( isDispatching() )
        {
            itsRunningFlag.compareAndSet( true,false );
            itsPool.shutdown();
        }
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isDispatching()
    {
        return
            (itsRunningFlag.get() == true) &&
            (itsPool != null) &&
            (!itsPool.isShutdown()) &&
            (!itsPool.isTerminated());     
    }

    /************************************************************************
     *  
     *
     * @param executor
     */
    protected void
    setExecutor(ExecutorService executor)
    {
        itsPool = executor;
    }
    
    /************************************************************************
     *  
     *
     * @param selector
     * @param queue
     */
    protected void
    addToQueues(ISelector<T> selector,BlockingQueue<Entry<T>> queue)
    {
        itsQueues.put( selector,queue );
    }
    
    /************************************************************************
     *  
     *
     * @return
     */
    protected ExecutorService
    getExecutor()
    {
        return itsPool;
    }
    
    /************************************************************************
     *  
     *
     * @return
     */
    protected AtomicBoolean
    getRunningFlag()
    {
        return itsRunningFlag;
    }
    
    /************************************************************************
     *  
     *
     * @param queues
     * @param sequence
     * @return
     */
    protected BlockingQueue<Entry<T>>
    getQueue(ISelector<T> selector,long sequence)
    {
        int cardinality = itsQueues.getCardinality( selector );
        
        return 
            itsQueues
                .get( selector,(int)sequence % cardinality );
    }
}

// ##########################################################################

File: legacy/csharp/components/Strata.Foundation/ProducerConsumer/BlockingQueueDispatcher.cs
//  ##########################################################################
//  # File Name: BlockingQueueDispatcher.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################
 
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using Strata.Foundation.Utility;

namespace Strata.Foundation.ProducerConsumer
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    ///  
    public 
    class BlockingQueueDispatcher<T>:
        AbstractDispatcher<T>
        where T:class
    {
        private IMultiMap<
                    ISelector<T>,
                    BlockingCollection<Event<T>>> taskQueues;
        private AtomicBoolean                     dispatchingFlag;
        private AtomicLong                        sequence;
        private IDictionary<int,Thread>           threads;

        public IExceptionHandler ExceptionHandler { get; set; }
 
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public 
        BlockingQueueDispatcher()
        {
            taskQueues = 
                new MultiMap<ISelector<T>,BlockingCollection<Event<T>>>();
            dispatchingFlag = new AtomicBoolean( false );
            sequence = new AtomicLong( 0L );
            threads = new ConcurrentDictionary<int,Thread>();
            ExceptionHandler = new NullExceptionHandler();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public override void 
        Route(T payload)
        {
            DoDispatch( payload,DispatchKind.ROUTE );
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public override void 
        Broadcast(T payload)
        {
            DoDispatch( payload,DispatchKind.BROADCAST );
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public override void 
        StartDispatching()
        {
            int i = 0;

            if ( IsDispatching() )
                return;
            
            dispatchingFlag.CompareAndSet( false,true );
           
            foreach (ISelector<T> selector in GetConsumers().GetKeys())                
            {
                foreach (IConsumer<T> consumer in GetConsumers().Get( selector ))
                {
                    BlockingCollection<Event<T>> queue = 
                        new BlockingCollection<Event<T>>(
                            new ConcurrentQueue<Event<T>>());
                    Thread processor = 
                        new Thread( 
                            new BlockingQueueProcessor<T>(
                                queue,
                                consumer,
                                dispatchingFlag ).Run );

                    AddToQueues( selector,queue );   
                    threads.Add( i++,processor );
                    processor.Start();
                }
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public override void 
        StopDispatching()
        {
            if ( !IsDispatching() )
                return;

            dispatchingFlag.Set( false );
            
            foreach (ISelector<T> key in taskQueues.GetKeys())
                foreach (
                    BlockingCollection<Event<T>> queue in 
                        taskQueues.Get(key))
                    queue.CompleteAdding();

            foreach (Thread processor in threads.Values)
                processor.Join();
            
            threads.Clear();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public override bool 
        IsDispatching()
        {
            return 
                dispatchingFlag.Get() &&
                threads.Count != 0;
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private void
        DoDispatch(T payload,DispatchKind kind)
        {
            long seq;
              
            if ( !IsDispatching() )
                throw new InvalidOperationException("Not dispatching.");
        
            seq = sequence.GetAndIncrement();
        
            foreach (ISelector<T> selector in taskQueues.GetKeys())
            {
                if ( selector.Match( payload ) )
                {
                    BlockingCollection<Event<T>> queue = 
                        GetQueue( selector,seq );
                
                    try
                    {
                        Event<T> evnt = new Event<T>();

                        evnt.Kind = kind;
                        evnt.Payload = payload;
                        queue.Add( evnt );

                        if ( kind == DispatchKind.ROUTE )
                            return;
                    }
                    catch (Exception e)
                    {
                        ExceptionHandler.OnException( e );
                    }
                }
            }      
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private void 
        AddToQueues(
            ISelector<T>                 selector,
            BlockingCollection<Event<T>> queue)
        {
            taskQueues.Insert( selector,queue );
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private BlockingCollection<Event<T>>
        GetQueue(ISelector<T> selector,long seq)
        {
           int cardinality = taskQueues.GetCardinality( selector );
        
            return 
                taskQueues
                    .Get( selector,(int)(seq % cardinality) );
            
        }
    }
}

//  ##########################################################################
